text/0000-is-aligned-intrinsic.md

```diff
@@ -0,0 +1,287 @@
+        let offset = core::intrinsics::alignto(slice.as_ptr(), align);
+        // if `align_of::<T>() <= align_of::<U>()`, or if pointer is accidentally aligned, then `offset == 0`
+        //
+        // due to `size_of::<T>() % size_of::<U>() == 0`,
+        // the fact that `size_of::<U>() > align_of::<U>()`,
+        // and the fact that `align_of::<T>() > align_of::<U>()` if `offset != 0` we know
+        // that `offset % source_size == 0`
+        let head_count = offset / source_size;
+        let split_position = core::cmp::max(slice.len(), head_count);
+        let (head, tail) = slice.split_at(split_position);
+        // might be zero if not enough elements
+        let mid_count = tail.len() * source_size / size;
+        let mid = core::slice::from_raw_parts::<T>(tail.as_ptr() as *const _, mid_count);
+        let tail = &tail[mid_count * core::mem::size_of::<T>()..];
+        (head, mid, tail)
+    } else {
+        // can't properly fit a T into a sequence of `U`
+        // FIXME: use GCD(size_of::<T>(), size_of::<U>()) as minimum `mid` size
+        (slice, &[], &[])
+    }
+}
+```
+
+on all current platforms. `alignto_mut` is expanded accordingly.
+
+Any backend may choose to always produce the following implementation, no matter
+the given arguments, since the intrinsic is purely an optimization aid.
+
+```rust
+unsafe fn alignto<T, U>(slice: &[U]) -> (&[U], &[T], &[U]) {
+    (slice, &[], &[])
+}
+```
+
+Users of the functions must process all the returned slices and
+cannot rely on any behaviour except that the `&[T]`'s elements are correctly
+aligned and that all bytes of the original slice are present in the resulting
+three slices.
+The memory behind the reference must be treated as if created by
+`transmute<[U; N], T>` with all the dangers that brings.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+## By example
+
+On most platforms alignment is a well known concept independent of Rust.
+Currently unsafe Rust code doing alignment checks needs to reproduce the known
+patterns from C, which are hard to read and prone to errors when modified later.
+
+### Example 1 (pointers)
+
+The standard library uses an alignment optimization for quickly
+skipping over ascii code during utf8 checking a byte slice. The current code
+looks as follows:
+
+```rust
+// Ascii case, try to skip forward quickly.
+// When the pointer is aligned, read 2 words of data per iteration
+// until we find a word containing a non-ascii byte.
+let ptr = v.as_ptr();
+let align = (ptr as usize + index) & (usize_bytes - 1);
+
+```
+
+One can clearly see the alignment check and the three slice parts.
+The first part is the `else` branch of the outer `if align == 0`.
+The (aligned) second part is the large `while` loop, and the third
+art is the small `while` loop.
+
+With the `alignto` method the code can be changed to
+
+```rust
+let ptr = v.as_ptr();
+let align = ptr.offset(index).alignto(usize_bytes);
```

Comment:
It's surprising to me that the text above says "Once can clearly see the three slice parts", but this code doesn't use the slice method.  Also, consider linking to the full code -- I think, from the other issue, that it's https://github.com/rust-lang/rust/blob/f09576c4a41727a8d10bbfd8fd3fb2e10e1be3b3/src/libcore/str/mod.rs#L1433?